packages/liveblocks-react-ui/src/primitives/AiMessage/index.tsx

```diff
@@ -36,6 +36,15 @@ const defaultMessageContentComponents: AiMessageContentComponents = {
  * ToolInvocationPart
  * -----------------------------------------------------------------------------------------------*/
 
+function CustomTool(props: {
+  renderFn: FunctionComponent<
+    AiToolInvocationProps<JsonObject, ToolResultData>
+  >;
+  props: AiToolInvocationProps<JsonObject, ToolResultData>;
+}) {
+  return props.renderFn(props.props);
+}
```

Comment:
The final trick is this:

- Rather than invoking `<tool.render />`, instead we invoke `<CustomTool renderFn={tool.render} />`.
- This gives React a stable reference (the value of `CustomTool` is constant), and thus a place to associate state in the tree.
- The implementation will still "just" call `tool.render(props)`, returning the JSX.